orchagent/pfc_detect_broadcom.lua

```diff
@@ -0,0 +1,91 @@
+        end
+
+        local queue_index = redis.call('HGET', 'COUNTERS_QUEUE_INDEX_MAP', KEYS[i])
+        local port_id = redis.call('HGET', 'COUNTERS_QUEUE_PORT_MAP', KEYS[i])
+        local pfc_rx_pkt_key = 'SAI_PORT_STAT_PFC_' .. queue_index .. '_RX_PKTS'
+        local pfc_on2off_key = 'SAI_PORT_STAT_PFC_' .. queue_index .. '_ON2OFF_RX_PKTS'
+        
+
+        -- Get all counters
+        local occupancy_bytes = tonumber(redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_QUEUE_STAT_CURR_OCCUPANCY_BYTES'))
+        local packets = tonumber(redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_QUEUE_STAT_PACKETS'))
+        local pfc_rx_packets = tonumber(redis.call('HGET', counters_table_name .. ':' .. port_id, pfc_rx_pkt_key))
+        local pfc_on2off = tonumber(redis.call('HGET', counters_table_name .. ':' .. port_id, pfc_on2off_key))
+        local queue_pause_status = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_QUEUE_ATTR_PAUSE_STATUS')
+        
+        local packets_last = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_QUEUE_STAT_PACKETS_last')
+        local pfc_rx_packets_last = redis.call('HGET', counters_table_name .. ':' .. port_id, pfc_rx_pkt_key .. '_last')
+        local pfc_on2off_last = redis.call('HGET', counters_table_name .. ':' .. port_id, pfc_on2off_key .. '_last')
+        local queue_pause_status_last = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'SAI_QUEUE_ATTR_PAUSE_STATUS_last')
+
+        -- DEBUG CODE START. Uncomment to enable
+        local debug_storm = redis.call('HGET', counters_table_name .. ':' .. KEYS[i], 'DEBUG_STORM')
+        -- DEBUG CODE END.
+
+        -- If this is not a first run, then we have last values available
+        if packets_last and pfc_rx_packets_last and pfc_on2off_last and queue_pause_status_last then
+            packets_last = tonumber(packets_last)
+            pfc_rx_packets_last = tonumber(pfc_rx_packets_last)
+            pfc_on2off_last = tonumber(pfc_on2off_last)
+
+            -- Check actual condition of queue being in PFC storm
+            if (occupancy_bytes > 0 and packets - packets_last == 0 and pfc_rx_packets - pfc_rx_packets_last > 0) or
+                -- DEBUG CODE START. Uncomment to enable
+                (debug_storm == "enabled") or
+                -- DEBUG CODE END.
+                (occupancy_bytes == 0 and pfc_rx_packets - pfc_rx_packets_last > 0 and pfc_on2off - pfc_on2off_last == 0 and queue_pause_status_last == 'true' and queue_pause_status == 'true') then
+                if time_left <= poll_time then
+                    redis.call('PUBLISH', 'PFC_WD', '["' .. KEYS[i] .. '","storm"]')
+                    is_deadlock = true
+                    time_left = detection_time
+                else
+                    time_left = time_left - poll_time
+                end
+            else
+                if pfc_wd_action == 'alert' and pfc_wd_status ~= 'operational' then
```

Comment:
Plugins do not distinguish between actions. Two types of notifications are "storm" and "restore". Orchagent decides what to do based on configuration.

Reply:
Currently, the detect plugin checks queues in operational status, and restore plugin checks queues in stormed status. This is not true for the alert mode. In alert mode, the detect plugin will check all the queues regardless of the queue status, and restore plugin never check anything. Currently the logic of when/which queue to check is included in the plugin itself. That's the reason why plugin also needs to know the action  type to decide which queue to check for alert-mode and normal case. In https://github.com/Azure/sonic-swss/pull/342, it still runs restore plugins for stormed queues in alert mode, which is not correct.

Reply:
According to the document, definition of alert mode is following:
PFC watchdog logs when the action of dropping packets when pause storm occurs and resume forwarding traffic when pause storm stops. With the alert mode, PFC watchdog only need to log that pause storm detected and pause storm stopped.

So as I understand, alert mode is same as other actions, except it doesn't apply any configuration, but only logs events.

@lguohan can you please take a look? Maybe I'm missing something?

Reply:
The restore criteria is that we don't receive PFC any more. This actually bases on the assumption that we don't send any packets because of the recovery action we have taken. In alert mode, we are not doing anything but log the events. In such case, restore criteria is too strong as it's OK to occasionally receive PFC in normal case. That is also the reason why we need to use the detection criteria to see whether the storm is stopped in such case.

Reply:
ok